eppo_client/bandit.py

```diff
@@ -0,0 +1,271 @@
+        weights = [
+            (
+                action_key,
+                max(
+                    min_probability,
+                    1.0 / (number_of_actions + gamma * (best_score - score)),
+                ),
+            )
+            for action_key, score in action_scores
+            if action_key != best_action
+        ]
+
+        # remaining weight goes to best action
+        remaining_weight = 1.0 - sum(weight for _, weight in weights)
+        weights.append((best_action, remaining_weight))
+        return weights
+
+    def select_action(self, flag_key, subject_key, action_weights) -> Tuple[int, str]:
+        # deterministic ordering
+        sorted_action_weights = sorted(
+            action_weights,
+            key=lambda t: self.sharder.get_shard(
+                f"{flag_key}-{subject_key}-{t[0]}", self.total_shards
+            ),
+        )
+
+        # select action based on weights
+        shard = self.sharder.get_shard(f"{flag_key}-{subject_key}", self.total_shards)
+        cumulative_weight = 0.0
+        shard_value = shard / self.total_shards
+
+        for idx, (action_key, weight) in enumerate(sorted_action_weights):
+            cumulative_weight += weight
+            if cumulative_weight > shard_value:
+                return idx, action_key
+
+        # If no action is selected, return the last action (fallback)
+        action_index = len(sorted_action_weights) - 1
+        action_key = sorted_action_weights[action_index][0]
+        return action_index, action_key
+
+
+def score_action(
+    subject_attributes: Attributes,
+    action_attributes: Attributes,
+    coefficients: BanditCoefficients,
+) -> float:
+    score = coefficients.intercept
+    score += score_numeric_attributes(
+        coefficients.subject_numeric_coefficients,
+        subject_attributes.numeric_attributes,
+    )
+    score += score_categorical_attributes(
+        coefficients.subject_categorical_coefficients,
+        subject_attributes.categorical_attributes,
+    )
+    score += score_numeric_attributes(
+        coefficients.action_numeric_coefficients,
+        action_attributes.numeric_attributes,
+    )
+    score += score_categorical_attributes(
+        coefficients.action_categorical_coefficients,
+        action_attributes.categorical_attributes,
+    )
+    return score
+
+
+def coalesce(value, default=0):
+    return value if value is not None else default
+
+
+def score_numeric_attributes(
+    coefficients: List[BanditNumericAttributeCoefficient],
+    attributes: Dict[str, float],
+) -> float:
+    score = 0.0
+    for coefficient in coefficients:
+        print(coefficient, attributes)
+        if coefficient.attribute_key in attributes:
+            score += coefficient.coefficient * coalesce(
+                attributes[coefficient.attribute_key],
+                coefficient.missing_value_coefficient,
+            )
+        else:
+            score += coefficient.missing_value_coefficient
```

Comment:
```suggestion
        if coefficient.attribute_key in attributes and attributes[coefficient.attribute_key] is not None:
            score += coefficient.coefficient * attributes[coefficient.attribute_key]
        else:
            score += coefficient.missing_value_coefficient
```
The `coalesce` surprises me; why would we multiply the (proper) coefficient by the missing value coefficient? If `attributes[coefficient.attribute_key]` is null, then we should fall into the else case. See suggested edit.